helpers/dbload.ts

```diff
@@ -96,7 +96,8 @@ User.hasMany(UserLesson, {
 })
 
 UserLesson.belongsTo(User, {
-  foreignKey: 'userId'
+  foreignKey: 'userId',
```

Comment:
Does this mean the test database is different than the production database? If that is the case, that should be fixed too.

Reply:
No, the database is unaltered because I didn't change the foreign key.

The `as` value is just the internal name that Sequelize will use when defining the associations. By default it uses the model name, but since the build step changes that we need to set an alias to make sure the property name is the same in dev and prod. ([Sequelize guide](https://sequelize.org/v5/manual/associations.html#naming-strategy), [API reference](https://sequelize.org/v5/class/lib/model.js~Model.html#static-method-belongsTo))

Reply:
I understand what happens in sequelize. I am concerned about why this works in development but not in production.

Reply:
It breaks in prod because webpack changes the model names so `User` becomes `User_User`.

If you run `yarn build` and look at the transpiled files you'll see this for the `User` model definition:

```js
class User_User extends external_sequelize_["Model"] {}
```

And the `getLessonMentors` gets transpiled to this:

```js
const getLessonMentors = async (_parent, args) => {
  const {
    lessonId
  } = args;

  try {
    await validateLessonId(parseInt(lessonId));
    const results = await dbload["b" /* UserLesson */].findAll({
      where: {
        lessonId
      },
      include: [{
        model: dbload["a" /* User */]
      }]
    });
    return results.map(({
      User: {
        username,
        name,
        id
      }
    }) => {
      return {
        username,
        name,
        id
      };
    });
  } catch (err) {
    throw new Error(err);
  }
};
```

The part that breaks in prod in that the `User` field doesn't exist because it changed to `User_User`.

Reply:
This will work in kinda of a improvised fix until we can find a better way to make this work more easily.

Reply:
Since `sequelize` is backend code, it shouldn't be transpired by webpack. What you see above looks like it's code that is generated by `graphql-codegen` 

This tells me that our development database may have a property called `User` in `user`. We should check that and fix this. We possibly may  want to want `alter` in development to correct this.

Reply:
`graphql-codegen` generates only one file: [`graphql/index.tsx`](https://github.com/garageScript/c0d3-app/blob/master/graphql/index.tsx), and that file only has the types for the resolver ([1](https://github.com/garageScript/c0d3-app/blob/70c5ccbed18bae71f7604e64ca4aaccf9351a5cf/graphql/index.tsx#L198), [2](https://github.com/garageScript/c0d3-app/blob/70c5ccbed18bae71f7604e64ca4aaccf9351a5cf/graphql/index.tsx#L205), [3](https://github.com/garageScript/c0d3-app/blob/70c5ccbed18bae71f7604e64ca4aaccf9351a5cf/graphql/index.tsx#L509), [4](https://github.com/garageScript/c0d3-app/blob/70c5ccbed18bae71f7604e64ca4aaccf9351a5cf/graphql/index.tsx#L1151)). And besides that, it only generates that file if you run the command for it, it's not a build step. Also the graphql types and sequelize model definitions are independent of each other right now.

The code above is generated after running `yarn build` (which is [aliased](https://github.com/garageScript/c0d3-app/blob/70c5ccbed18bae71f7604e64ca4aaccf9351a5cf/package.json#L6) to `next build`) like I already said before. And the code above is from the production-ready files inside the `.next` folder. That specific code is in the file `.next/server/pages/api/graphql.js`, that is generated from [this](https://github.com/garageScript/c0d3-app/blob/master/pages/api/graphql.ts) file. You can run `yarn build` and see for yourself. That is the code that is running on the deployed version on Vercel right now.

Even if it's backend code it's still managed by the Next.js framework, because that is how Next.js [works](https://nextjs.org/docs/api-routes/introduction). I don't know if it's webpack or babel that makes the code like that, but it's not `graphql-codegen`. @Ulisseus also sent a related issue with next-auth and TypeORM, [see here](https://chat.c0d3.com/c0d3/pl/ecnzz4itubywxqe6k31c19zwxh).

The development database has nothing different as well, it's the same columns, I didn't even touch the database. The alias is just for sequelize to put the value in the right property on the resulting JS object. [There's even another aliases already in use](https://github.com/garageScript/c0d3-app/blob/70c5ccbed18bae71f7604e64ca4aaccf9351a5cf/helpers/dbload.ts#L67-L83).

Running the server with `next start` and attaching a debugger, these are the fields on one of the values of `results`:

```js
console.log(JSON.stringify(results[0]))
{
  "isPassed": "1613224106554",
  "isTeaching": "1613224113460",
  "isEnrolled": "1613224116772",
  "lessonId": 5,
  "createdAt": "2021-01-01T02:04:36.339Z",
  "updatedAt": "2021-01-01T11:57:00.101Z",
  "userId": 1,
  "User_User": {
    "id": 1,
    "name": "Name",
    "username": "username",
    "password": "redacted",
    "email": "user@email.com",
    "gsId": null,
    "isOnline": null,
    "isAdmin": "false",
    "forgotToken": null,
    "cliToken": null,
    "emailVerificationToken": null,
    "tokenExpiration": null,
    "createdAt": "2021-01-01T01:24:43.803Z",
    "updatedAt": "2021-01-01T11:46:18.082Z"
  }
}
```

And here is that same value but running with `next dev`:

```js
console.log(JSON.stringify(results[0]))
{
  "isPassed": "1613224106554",
  "isTeaching": "1613224113460",
  "isEnrolled": "1613224116772",
  "lessonId": 5,
  "createdAt": "2021-01-01T02:04:36.339Z",
  "updatedAt": "2021-01-01T11:57:00.101Z",
  "userId": 1,
  "User": {
    "id": 1,
    "name": "Name",
    "username": "username",
    "password": "redacted",
    "email": "user@email.com",
    "gsId": null,
    "isOnline": null,
    "isAdmin": "false",
    "forgotToken": null,
    "cliToken": null,
    "emailVerificationToken": null,
    "tokenExpiration": null,
    "createdAt": "2021-01-01T01:24:43.803Z",
    "updatedAt": "2021-01-01T11:46:18.082Z"
  }
}
```

So you see that the expected `User` field is not present in the production build, instead we have `User_User` because it's the model name that was changed in the files bundled by `next build`.

> This tells me that our development database may have a property called `User` in `user`.

Why would you think that?

Besides trying to fix this on our side, the only other viable solution I see would be to open a issue on the Next.js repo and see what it's developers have to say about that.

Reply:
It looks like it's babel running through webpack that transforms the code like this.

The **root cause** isn't with the backend code but just with now `Next.JS` wants to compile everything since it's running through the serverless function with `Vercel`.

I was simply worried that our dev environment has `userLesson` schema of below. 

```
{
  "isPassed": "1613224106554",
  "isTeaching": "1613224113460",
  "isEnrolled": "1613224116772",
  "lessonId": 5,
  "createdAt": "2021-01-01T02:04:36.339Z",
  "updatedAt": "2021-01-01T11:57:00.101Z",
  "userId": 1,
  "User_User": {
    "id": 1,
    "name": "Name",
    "username": "username",
    "password": "redacted",
    "email": "user@email.com",
    "gsId": null,
    "isOnline": null,
    "isAdmin": "false",
    "forgotToken": null,
    "cliToken": null,
    "emailVerificationToken": null,
    "tokenExpiration": null,
    "createdAt": "2021-01-01T01:24:43.803Z",
    "updatedAt": "2021-01-01T11:46:18.082Z"
  }
}
```

With your comments above and I also checked at @anthonykhoa's development server and it looks good. 

I didn't actually have any concerns with the PR itself, but I just wanted to make sure we're doing due diligence on managing the db side between prod and app. I also don't regard this as an improvised fix. We are simply being more explicit in what we're calling our relations in SQL.